server/src/main/java/org/opensearch/ingest/IngestService.java

```diff
@@ -571,13 +580,286 @@ protected void doRun() {
+        }
+        return batchedIndexRequests;
+    }
+
+    /* visible for testing */
+    static final class IndexRequestWrapper {
+        private final int slot;
+        private final IndexRequest indexRequest;
+        private final List<String> pipelines;
+        private final boolean hasFinalPipeline;
+
+        IndexRequestWrapper(int slot, IndexRequest indexRequest, List<String> pipelines, boolean hasFinalPipeline) {
+            this.slot = slot;
+            this.indexRequest = indexRequest;
+            this.pipelines = pipelines;
+            this.hasFinalPipeline = hasFinalPipeline;
+        }
+
+        public int getSlot() {
+            return slot;
+        }
+
+        public IndexRequest getIndexRequest() {
+            return indexRequest;
+        }
+
+        public List<String> getPipelines() {
+            return pipelines;
+        }
+
+        public boolean isHasFinalPipeline() {
+            return hasFinalPipeline;
+        }
+    }
+
+    private void executePipelinesInBatchRequests(
+        final List<Integer> slots,
+        final Iterator<String> pipelineIterator,
+        final boolean hasFinalPipeline,
+        final List<IndexRequest> indexRequests,
+        final IntConsumer onDropped,
+        final BiConsumer<Integer, Exception> onFailure,
+        final AtomicInteger counter,
+        final BiConsumer<Thread, Exception> onCompletion,
+        final Thread originalThread
+    ) {
+        if (indexRequests.size() == 1) {
+            executePipelines(
+                slots.get(0),
+                pipelineIterator,
+                hasFinalPipeline,
+                indexRequests.get(0),
+                onDropped,
+                onFailure,
+                counter,
+                onCompletion,
+                originalThread
+            );
+            return;
+        }
+        while (pipelineIterator.hasNext()) {
+            final String pipelineId = pipelineIterator.next();
+            try {
+                PipelineHolder holder = pipelines.get(pipelineId);
+                if (holder == null) {
+                    throw new IllegalArgumentException("pipeline with id [" + pipelineId + "] does not exist");
+                }
+                Pipeline pipeline = holder.pipeline;
+                String originalIndex = indexRequests.get(0).indices()[0];
+                Map<Integer, IndexRequest> slotIndexRequestMap = createSlotIndexRequestMap(slots, indexRequests);
+                innerBatchExecute(slots, indexRequests, pipeline, onDropped, results -> {
+                    for (int i = 0; i < results.size(); ++i) {
+                        if (results.get(i).getException() != null) {
+                            IndexRequest indexRequest = slotIndexRequestMap.get(results.get(i).getSlot());
+                            logger.debug(
+                                () -> new ParameterizedMessage(
+                                    "failed to execute pipeline [{}] for document [{}/{}]",
```

Comment:
We don't consistently capitalize these errors, do you know which one is more correct? "failed to ..." or "Failed to ... .".

Reply:
I do see more "failed to" in this repo. I haven't found a java specification about this but golang recommends not to capitalize [[source](https://go.dev/wiki/CodeReviewComments#error-strings)]. It's very interesting that I randomly chose Enum.java in JDK, and the exception messages have both.

I can dive deep to see if there is a good source of recommendation for java with good explanation. I'll leave the code here for now.